agent/lib/kontena/workers/service_pod_manager.rb

```diff
@@ -0,0 +1,126 @@
+require_relative 'service_pod_worker'
+require_relative '../models/service_pod'
+
+module Kontena::Workers
+  class ServicePodManager
+    include Celluloid
+    include Celluloid::Notifications
+    include Kontena::Logging
+    include Kontena::Helpers::RpcHelper
+
+    attr_reader :workers
+
+    trap_exit :on_worker_exit
+    finalizer :finalize
+
+    def initialize(autostart = true)
+      @workers = {}
+      subscribe('service_pod:update', :on_update_notify)
+      async.start if autostart
+    end
+
+    def start
+      node # blocks until it's available
+      populate_workers_from_docker
+      loop do
+        populate_workers_from_master
+        sleep 30
+      end
+    end
+
+    # @return [Node]
+    def node
+      if @node.nil?
+        @node = Actor[:node_info_worker].node
+      end
+
+      @node
+    end
+
+    def on_update_notify(_, _)
+      populate_workers_from_master
+    end
+
+    def populate_workers_from_master
+      exclusive {
+        request = rpc_client.request("/node_service_pods/list", [node.id])
+        response = request.value
+        return unless response['service_pods'].is_a?(Array)
+
+        service_pods = response['service_pods']
+        current_ids = service_pods.map { |p| p['id'] }
+        terminate_workers(current_ids)
+
+        service_pods.each do |s|
+          ensure_service_worker(Kontena::Models::ServicePod.new(s))
+        end
+      }
+    end
+
+    def populate_workers_from_docker
+      info "populating service pod workers from docker"
+      fetch_containers.each do |c|
+        service_pod = Kontena::Models::ServicePod.new(
+          'id' => "#{c.service_id}/#{c.instance_number}",
+          'service_id' => c.service_id,
+          'instance_number' => c.instance_number,
+          'desired_state' => "unknown"
```

Comment:
I do miss enums fron Java days... :)  Would it make sense to use some enum like solution for these states instead of semi-random strings? E.g. https://github.com/dblock/ruby-enum looks pretty ok.